### PR TITLE
Dev/timeline scrubbing

### PR DESCRIPTION
**Pull request by glabute**
Added optional timeline scrubbing cache feature.  
Description is here:
https://docs.google.com/document/d/1wb2V-VXPkgsWNgXb6jCF8wKtvGWY66JWaMjAC4y77ww/edit#heading=h.jxyw25p59zfr

### CODE REVIEW

File: Runtime/Core/TargetPositionCache.cs [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+using System.Collections.Generic;
+using System;
+
+namespace Cinemachine
+{
+    internal class TargetPositionCache
+    {
+        public static bool UseCache { get; set; }
+        public enum Mode { Disabled, Record, Playback }
+        public static Mode m_CacheMode = Mode.Disabled;
```
Comment (gaborkb) [inline Runtime/Core/TargetPositionCache.cs:11]: m_CacheMode could be private

File: Runtime/Core/TargetPositionCache.cs [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+                var ZList = new List<Keyframe>();
+                var RotXList = new List<Keyframe>();
+                var RotYList = new List<Keyframe>();
+                var RotZList = new List<Keyframe>();
+
+                RawItems.Sort();
+                float time = float.MaxValue;
+                for (int i = 0; i < RawItems.Count; ++i)
+                {
+                    var item = RawItems[i];
```
Comment (gaborkb) [inline Runtime/Core/TargetPositionCache.cs:77]: foreach (var item in RawItems)
{

Foreach may be nicer, but it's not necessarily better.

File: Runtime/Timeline/CinemachineMixer.cs [hunk 3]
```diff
@@ -6,26 +6,139 @@
...
+                    {
+                        var vcam = sublist[j];
+                        maxDampingTime = Mathf.Max(maxDampingTime, vcam.GetMaxDampTime());
+                    }
+                }
+                // Impose upper limit on damping time, to avoid simulating too many frames
+                return Mathf.Min(maxDampingTime, 2.0f); 
+            }
+
+            public void ScrubTohere(
```
Comment (gaborkb) [inline Runtime/Timeline/CinemachineMixer.cs:65]: ScrubToHere - casing typo

File: Runtime/Core/TargetPositionCache.cs (hunk 1 above)
Comment (glabute) [reply Runtime/Core/TargetPositionCache.cs:11]: yes

File: Runtime/Core/TargetPositionCache.cs (hunk 2 above)
Comment (glabute) [reply Runtime/Core/TargetPositionCache.cs:77]: I used to use foreach, but then discovered that there are situations where it does GC allocs.  So now I always do it this way

File: Runtime/Timeline/CinemachineMixer.cs (hunk 3 above)
Comment (glabute) [reply Runtime/Timeline/CinemachineMixer.cs:65]: yes

**Comment by glabute** [comment]
I'm worrying a little about performance.  I think AnimationCurve is very inefficient when it evaluates (it has to iterate all keys from beginning to find the desired time, every frame).  Maybe a custom implementation of a segmented linear with a fixed timestep would be better

File: Runtime/Core/TargetPositionCache.cs (hunk 2 above)
Comment (gaborkb) [reply Runtime/Core/TargetPositionCache.cs:78]: Thanks for the info! :)

**Pull request merged**